[RNMobile] Avoid keyboard dismiss when interacting text blocks

fluiddot opened the pull request:
**Related PRs:**
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/6467
- _**[For testing]**_  https://github.com/wordpress-mobile/WordPress-iOS/pull/22235
- _**[For testing]**_ https://github.com/wordpress-mobile/WordPress-Android/pull/19795

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

This PR solves the following cases that are affected by the keyboard’s disruptive hide-and-show behavior:
- Insert a new block by pressing the `enter` key in a text-based block.
- Merge blocks by pressing the `backspace` key in a text-based block.

However, there's a remaining case that this PR won't cover:
- Replace blocks when interacting with the List block.

We haven't found a solution to this yet, as a next step, we'd need to explore further the actions performed during block creation and merging in the List block.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

This PR will greatly improve the usability when writing content, as the hide-and-show behavior impacts negatively the editing experience.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

### Solve `blur`/`focus` event conflicts

Each `RichtText` component instance triggers `blur`/`focus` events based on its state (e.g. will be unmounted, gains/loses selection) without considering other instances states. This leads to conflicts between `blur` and `focus` events in different cases like pressing `enter` on a Paragraph block to insert a new paragraph.

Here is an example of the sequence of events generated when pressing `enter`:

- Paragraph 1 triggers `blur` event. This comes from the `componentDidUpdate` function due to changing its selection state from selected to unselected.
  - **This `blur` is the culprit of the keyboard being dismissed, as it executed first.**
- Paragraph 2 triggers `focus` event when it's mounted (i.e. from the `componentDidMount` function).
- **This `focus` event makes the keyboard to open.** 
- Paragraph 2 triggers `focus` a second time. However, this event is innocuous. This comes from the `componentDidUpdate` function due to changing its selection state from unselected to selected.
- Paragraph 1 triggers `blur` event. This comes from the  `_onBlur` callback triggered by Aztec because the text input lost focus. This event is innocuous.

To solve this conflict, this PR introduces the usage of `debounce` in the `blur` event (https://github.com/WordPress/gutenberg/pull/57070/commits/b6f237f1ad2bee2a4ab91182ac7920b8238abd2f). It will ensure that `focus` events take precedence when `blur` events have been triggered at the same time, as well as cancel them to avoid the keyboard to be dismissed.

### iOS - Keyboard dismisses automatically when text input is removed 

Another case to solve is when two blocks are merged. The editor internally replaces the blocks which implies removing one of them. On iOS, this presents a problem because the keyboard will be automatically dismissed. No `blur` events are generated so a different approach was taken.

The solution (https://github.com/WordPress/gutenberg/pull/57070/commits/a71b1c5df9e3cc9f3f9067cbb7e2d978f69b8919) implied tweaking the order of native commands invoked for `blur` and `focus` events ([reference](https://github.com/WordPress/gutenberg/blob/125c130818ebb83c50618722858e8315d90e76b8/packages/react-native-aztec/ios/RNTAztecView/RCTAztecViewManager.swift#L70-L82)). The `focus` event triggered by the new block created (i.e. the one that contains the merged content) happens at the same time the other block is removed. So, the workaround for this case is to ensure that `focus` events are prepended in the pending UI blocks list of the `UIManager` ([reference 1 ](https://github.com/facebook/react-native/blob/7ea7d946c643f076c29bcf11b927f7569e3c516f/React/Modules/RCTUIManager.m#L67), [reference 2](https://github.com/facebook/react-native/blob/7ea7d946c643f076c29bcf11b927f7569e3c516f/React/Modules/RCTUIManager.m#L508C16-L517)).

### Android - Avoid triggering `blur` events on unmounted instances

As mentioned below, when `RichText`/`AztecView` is unmounted a `blur` event is triggered. On Android, this behavior in combination with using `debounce` leads to an exception because we try to `blur` a text input that no longer exists. To address this, a new function has been introduced in the Aztec input state (`blurOnUnmount`) that is only called for the component unmounting case (https://github.com/WordPress/gutenberg/pull/57070/commits/634867f072bda9b6419c6c85162d5d2afdef386c). This function has two purposes:
1. Cancel potential previous `blur` calls to avoid exceptions.
2. Hide the keyboard, as it would happen in a regular `blur` event. Note that the call to hide the keyboard uses `debounce`. This is needed to avoid dismissing the keyboard if a `focus` event happens at the same time.

### Other changes

As a side note, the `blur` event triggered when unmounting the `RichText` component has been moved to the `AztecView` component as seems a better fit for this logic (https://github.com/WordPress/gutenberg/pull/57070/commits/d3fb48686426302c434d329676a1a1607e1fa528).

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

> [!NOTE]
> It's recommended to use the installable builds for testing:
> * [Android](https://github.com/wordpress-mobile/WordPress-Android/pull/19795#issuecomment-1857562794)
> * [iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/22235#issuecomment-1857565204)

### 1 - Post title

1. Open the app.
3. Create an empty post.
4. Observe that the title is focused.
5. Type some words and press enter.
6. Observe the following:
   1. The Paragraph block is focused.
   2. The keyboard remained open when transitioning.
7. Press backspace.
8. Observe the following depending on the platform:
   1. Android
      1. The Paragraph block remains focused.
      2. The keyboard remains open.
   2. iOS
      1. The Paragraph block remains focused.
      2. The keyboard reopens (this should be tackled in the future)

### 2 - Insert a new block by pressing the `enter` key in a text-based block

1. Open the app.
2. Create an empty post.
3. Select the first Paragraph block.
4. Type some words and press enter.
5.  Observe the following:
   1. A new Paragraph block is created and focused.
   2. The keyboard remained open when transitioning.

### 3 - Merge blocks by pressing the `backspace` key in a text-based block (iOS only)

1. Open the app.
2. Create an empty post.
3. Select the first Paragraph block.
4. Type some words and press enter.
6.  Press backspace.
7.  Observe the following:
    1.  The second Paragraph block is removed.
    2.  The first Paragraph block is focused and the text caret is at the end of the content.
    3.  The keyboard remained open when transitioning.
9.  Place the text caret between two words.
10.  Press enter.
11. Observe the following:
    1.  The paragraph is split into a new Paragraph block and it's focused.
    2.  The keyboard remained open when transitioning.
12. Press backspace.
13. Observe the following:
    1.  The paragraph is merged with the first one and it's focused.
    2.  The keyboard remained open when transitioning.

### 4 - Automatically dismiss the keyboard when selecting non-text-based-blocks

1. Open the app.
2. Create an empty post.
3.  Add an Image block and set an image.
4.  Tap on the caption field.
5.  Observe the following:
    1.  The caption field is focused.
    2.  The keyboard opens.
6.  Type some words and press enter.
7.  Observe the following:
    1.  A new Paragraph block is created and focused.
    2.  The keyboard remained open when transitioning.
8.  Press backspace.
    1.  The Image block is selected.
    2.  The keyboard closes.
9. Insert a Paragraph block and type some words.
10. Observe that the Paragraph block is focused.
11. Tap on the Image block.
12. Observe the following:
    1.  The Paragraph block is no longer focused.
    2.  The keyboard closes.

### 5 - Manually dismiss the keyboard

1. Open the app.
2. Create an empty post.
3.  Select the first Paragraph block.
4.  Observe the following:
    1.  The Paragraph block is focused.
    2.  The keyboard opens.
5.  Tap on the keyboard icon located in the toolbar.
6.  Observe the following:
    1.  The Paragraph block is no longer focused.
    2.  The keyboard closes.

### 6 - The keyboard is dismissed and restored when opening modals

**NOTE:** On Android, this case will be covered by https://github.com/WordPress/gutenberg/pull/57069.

1. Open the app.
2. Create an empty post.
3.  Select the first Paragraph block.
4.  Open the block settings.
5.  Observe the following:
    1.  The Paragraph block is no longer focused.
    2.  The block settings bottom sheet opens.
    3.  The keyboard closes.
6.  Dismiss the bottom sheet.
7.  Observe the following:
    1.  The Paragraph block is focused again.
    2.  The keyboard opens.
8.  Navigate to the bottom of the post.
9.  Tap on the empty area below the content.
10. Observe the following:
    1.  A new Paragraph block is created and focused.
    2.  The keyboard remained open when transitioning.

### 6 - Writing flow

It's recommended to cover the [writing flow test suite](https://github.com/wordpress-mobile/test-cases/tree/trunk/test-cases/gutenberg/writing-flow) by manually testing.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

## Screenshots or screencast <!-- if applicable -->

| Header | Header |
|--------|--------|
| <video src=https://github.com/WordPress/gutenberg/assets/14905380/7549afd9-a880-4436-b6cd-5380d8f7c010> | <video src=https://github.com/WordPress/gutenberg/assets/14905380/9afa6d86-b239-42a9-a526-f34f8b89db7c> |

fluiddot committed:
Use debounce in Aztec's blur function

fluiddot committed:
Execute `focus` UI block before other blocks

fluiddot committed:
Add `hideAndroidSoftKeyboard` to RN bridge

fluiddot committed:
Add `blurOnUnmount` to Aztec input state manager.

This function will help us to deal with the special case of unfocusing an Aztec view upon unmounting.

fluiddot committed:
Dismiss keyboard when Aztec view unmounts

This was previously handled in the `RichText` component.

fluiddot committed:
Fix unit test related to `AztecInputState` after adding debounce to `blur` function

fluiddot committed:
Remove console warning from `hideAndroidSoftKeyboard`

fluiddot committed:
Update inline comments of `blurOnUnmount` function

geriux reviewed (approved):
LGTM! This is a major improvement to the writing flow experience 🤩 

The code changes look good to me, I'm happy that we were able to achieve this with not that many code changes or refactors 🚀 

I've tested all test cases scenarios including the writing flow without any issues or regressions.

Devices used:
- iPhone 15 Pro (OS 17.1.2)
- Samsung Galaxy A23 5G (OS Android 13)
- Redmi Note 8T (OS Android 11)

Really nice work! 👏

fluiddot commented:
@geriux since this PR is already approved, let me know if you'd like to put the merge on hold until we merge https://github.com/WordPress/gutenberg/pull/57069 into this branch. From my side, this option makes sense.

geriux commented:
> @geriux since this PR is already approved, let me know if you'd like to put the merge on hold until we merge #57069 into this branch. From my side, this option makes sense as that PR is also approved.

I think to simplify things we could merge https://github.com/WordPress/gutenberg/pull/57069 into this branch and then merge this one into `trunk`

Commit:
[Mobile] - Android - Bring the Keyboard back when closing modals (#57069)

* React Native Bridge - Android - Introduces showAndroidSoftKeyboard to show the keyboard if there's a focused TextInput

* Mobile - Bottom Sheet - Adds usage of showAndroidSoftKeyboard when closing the Modal so it shows the Keyboard on Android for focused TextInputs

* React Native Bridge - Android - Introduces hideAndroidSoftKeyboard to hide the keyboard without triggering blur events

* React Native Bridge - Remove console warnings for unsupported methods, as their names are self-explanatory.

* Update showAndroidSoftKeyboard to take into account when the window focus changed, when we show the Modals these are shown on top of the editor activity.

It also adds an option to delay this for full screen modals

* Mobile - BottomSheet - Enable hardwareAccelerated and useNativeDriverForBackdrop props to improve performance on Android

* Update snapshots

* Removes hasWindowFocus condition as it is not being called hence not needed

* Refactor showAndroidSoftKeyboard to split into several functions, it also removes the delay functionality as it is no longer needed. It fixes an issue where mKeyboardRunnable was not being set.

It removes the delay logic from the Bottom Sheet component and the bridge.

* Updates createShowKeyboardRunnable to get the activity within the runnable instead of getting it as an param

* Remove unneeded check

geriux commented:
The changes from https://github.com/WordPress/gutenberg/pull/57069 are now merged into this branch 🎉

fluiddot commented:
Before heading to merge the PR, I'll update the changelog to reflect the fixes.

fluiddot committed:
Update `react-native-editor` changelog

jhnstn commented:
Hi @fluiddot, I ran into a crash on Android from these changes. 
I get the following error when trying to insert a new block from inserter menu: `Unable to execute operation dispatchViewManagerCommand: blur on view with tag: 1633, since the view does not exist`
It doesn't seem limited to a specific block but I am able to get new paragraph blocks when hitting return from another rich text block (post title or initial paragraph).

Screen:

https://github.com/WordPress/gutenberg/assets/744755/48210225-9d83-4d3c-a1fb-da4dd071f5c5

I get the error from anywhere on the post with or without any existing content.

I'm running the app locally on a Pixel 8 with WP-Android checked out at  https://github.com/wordpress-mobile/WordPress-Android/commit/70eb2d9f62b6bf2f19a0ced8bcb836c524e6d072 . I don't get the error after reverting the PR merge commit.

geriux commented:
Hey @jhnstn 👋 I'll investigate this issue, thanks for reporting!

geriux commented:
@jhnstn This issue is happening because we haven't released an alpha version of Gutenberg Mobile and these changes include native code.

If you build Gutenberg Mobile from source it should fix it, you can do this by changing your local `local-builds.gradle` file in WordPress Android and uncommenting the `localGutenbergMobilePath = "../gutenberg-mobile"` line.

Let us know if this fixes the problem until we create an alpha tag.

jhnstn commented:
Thanks @geriux !  setting up the local GB path fixed it
